mempool/cat/pool.go

```diff
@@ -0,0 +1,733 @@
+	)
+	checkTxResp := &abci.ResponseCheckTx{
+		Code:      abci.CodeTypeOK,
+		Priority:  info.priority,
+		Sender:    info.sender,
+		GasWanted: info.gasWanted,
+	}
+	return txmp.addNewTransaction(wtx, checkTxResp)
+}
+
+// CheckTx adds the given transaction to the mempool if it fits and passes the
+// application's ABCI CheckTx method. This should be viewed as the entry method for new transactions
+// into the network. In practice this happens via an RPC endpoint
+func (txmp *TxPool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo mempool.TxInfo) error {
+	// Reject transactions in excess of the configured maximum transaction size.
+	if len(tx) > txmp.config.MaxTxBytes {
+		return mempool.ErrTxTooLarge{Max: txmp.config.MaxTxBytes, Actual: len(tx)}
+	}
+
+	// This is a new transaction that we haven't seen before. Verify it against the app and attempt
+	// to add it to the transaction pool.
+	key := tx.Key()
+	rsp, err := txmp.TryAddNewTx(tx, key, txInfo)
+	if err != nil {
+		return err
+	}
+
+	// push to the broadcast queue that a new transaction is ready
+	txmp.markToBeBroadcast(key)
+
+	// call the callback if it is set
+	if cb != nil {
+		cb(&abci.Response{Value: &abci.Response_CheckTx{CheckTx: rsp}})
+	}
+	return nil
+}
+
+func (txmp *TxPool) next() <-chan types.TxKey {
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	if len(txmp.txsToBeBroadcast) != 0 {
+		ch := make(chan types.TxKey, 1)
+		for key := range txmp.txsToBeBroadcast {
+			delete(txmp.txsToBeBroadcast, key)
+			ch <- key
+			return ch
+		}
+	}
+	return txmp.broadcastCh
+}
+
+func (txmp *TxPool) markToBeBroadcast(key types.TxKey) {
+	if !txmp.config.Broadcast {
+		return
+	}
+
+	select {
+	case txmp.broadcastCh <- key:
+	default:
+		txmp.broadcastMtx.Lock()
+		defer txmp.broadcastMtx.Unlock()
+		txmp.txsToBeBroadcast[key] = struct{}{}
+	}
+}
+
+// TryAddNewTx attempts to add a tx that has not already been seen before. It first marks it as seen
+// to avoid races with the same tx. It then call `CheckTx` so that the application can validate it.
+// If it passes `CheckTx`, the new transaction is added to the mempool solong as it has
+// sufficient priority and space else if evicted it will return an error
+func (txmp *TxPool) TryAddNewTx(tx types.Tx, key types.TxKey, txInfo mempool.TxInfo) (*abci.ResponseCheckTx, error) {
+	// First check any of the caches to see if we can conclude early. We may have already seen and processed
+	// the transaction if:
+	// - We are conencted to legacy nodes which simply flood the network
+	// - If a client submits a transaction to multiple nodes (via RPC)
+	// - We send multiple requests and the first peer eventually responds after the second peer has already
+	if txmp.IsRejectedTx(key) {
+		// The peer has sent us a transaction that we have previously marked as invalid. Since `CheckTx` can
+		// be non-deterministic, we don't punish the peer but instead just ignore the msg
```

Comment:
[nit] I think the mempool only deals with transactions and doesn't inspect the contents (i.e. messages)

```suggestion
		// be non-deterministic, we don't punish the peer but instead just ignore the tx
```